---
title: "TD GWAS - Practical Session"
date: "2020-11-06"
author: "Lijiao Ning"
output: 
  html_document: 
    toc: true
    toc_depth: 3
    toc_float: 
      collapsed: false
    fig_width: 6.3
    fig_height: 4.7
    number_sections: true
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
  echo = TRUE, 
  cache.path = "figs"
)
```

# Practical Session Goals

* Know key steps of GWAS data quality control (QC) and the principles behind
* Know how to use PLINK to carry out QC steps
* Know how to run association test via PLINK and interpret the results
* Bonus: use the R-package ggplot2 for data visualization

# About The Data

This practical session uses a subset of the freely [HapMap](https://www.sanger.ac.uk/resources/downloads/human/hapmap3.html) data. The original HapMap data can be found here: ftp://ftp.ncbi.nlm.nih.gov/hapmap/genotypes/2010-05_phaseIII/plink_format/. 

In order to create a homogeneous ethnic population and to speed up the data processing steps, only European samples (the CEU population) were selected. 

A binary trait was simulated only for founders and was added to the dataset. 

We refer hereafter this dataset as "hapmap_ceu".

# Dowaload Materials

```bash
### build a directory for this TD
mkdir TD_GWAS
cd TD_GWAS

### download PLINK 1.9
wget http://s3.amazonaws.com/plink1-assets/plink_linux_x86_64_20201019.zip
unzip plink_linux_x86_64_20201019.zip
rm plink_linux_x86_64_20201019.zip

## For Mac http://s3.amazonaws.com/plink1-assets/plink_mac_20201019.zip 
## For Windows http://s3.amazonaws.com/plink1-assets/plink_win64_20201019.zip

### download "hapmap_ceu" data and Rmardown file
wget https://share-good.egid.fr/fop/VZfxQvAD/TD_GWAS_data.zip
unzip TD_GWAS_data.zip
rm TD_GWAS_data.zip
```

Please change directory to `TD_GWAS` if you are not already in the folder. Open the `TD_GWAS.Rmd` with Rstudio. Make sure the current working directory in Rstudio is `TD_GWAS`:

```{r}
getwd()
```

Otherwise, use the R function `setwd("path/to/your/TD_GWAS")` to change the working directory in Rstudio.

Now we have everything in the same directory, let's start.

# Describing Data

```{r}
## load packages
suppressMessages(library(dplyr))
suppressMessages(library(tidyr))
# suppressMessages(library(scales))
# suppressMessages(library(data.table))
suppressMessages(library(ggplot2))
theme_set(theme_light())

## source functions
source("viz_functions.R") # R functions for visualization
```

**Note**: in order to execute directly system command via Rstudio, we'll use the R function `system(command = "")`. Here, system command should be enclosed by a pair of quotes. Type `?system` in R console to see the detail description about this function. 
For Linux and Mac OS users, you can copy paste the command between quotes to run it in terminal. 
For Windows user, the command may not work in Windows terminal and need to be adapted to Windows command, if you are not familiar with Windows terminal command, please execute all system commands via Rstudio with `system()`.

**Quiz** 

* How many SNPs?

```{r}
system(command = "wc -l hapmap_ceu.bim", intern = TRUE)
```

* How many individuals? How many males/females? How many cases/controls?

```{r}
fam <- data.table::fread("hapmap_ceu.fam", header = FALSE)
table("sex" = fam$V5, "status" = fam$V6)
```

(Clue: [PLINK file format](https://www.cog-genomics.org/plink/2.0/formats))

# Quality Control

## Sample-based QC

### Sample Call Rate

The PLINK command `--missing` calculates the missing rate, here we will check the call rate as `1 - missing rate`.

```{r sample-cr-cmd}
system(command = "./plink --bfile hapmap_ceu --missing --out hapmap_ceu")
```

```{r sample-cr}
sample_callrate_thres <- 0.99 # the threshold of sample call rate, can be changed

sample_miss <- data.table::fread(
  file = "hapmap_ceu.imiss", 
  header = TRUE
) %>% 
  mutate(
    call_rate = 1 - F_MISS, 
    lab = ifelse(call_rate < sample_callrate_thres, IID, NA) # label bad call rate sample
  )

arrange(sample_miss, call_rate) %>% 
  ggplot(aes(x = seq_along(IID), y = call_rate)) +
  geom_point(shape = 21) +
  geom_hline(yintercept = sample_callrate_thres, color = "blue", linetype = 2) +
  geom_text(aes(label = lab), size = 3, hjust = -0.2, na.rm = TRUE) +
  scale_y_continuous(labels = scales::percent_format(accuracy = 0.1)) +
  labs(x = "Samples", y = "Call Rate", title = "Sample Call Rate Check")

## build a list to store samples to be excluded in each QC steps
sample_to_exclude <- list(
  "sample_cr" = filter(sample_miss, !is.na(lab)) %>% 
    select(FID, IID) %>% 
    mutate(QC = "Call rate")
)
```

**Quiz** 

* What's the possible reason when samples have a lot of missing genotypes?

=> *e.g.*: low DNA quality, sequencing depth is the same for every samples

### Heterozygosity Check

The observed heterozygosity rate per sample was computed using the formula
(Number of non-missing genotypes N(NM) - Number of homozygous genotypes O(Hom)) / N(NM).

N(NM) and O(Hom) were obtained using the `--het` command in PLINK.

```{r sample-het-cmd}
system(command = "./plink --bfile hapmap_ceu --het --out hapmap_ceu")
```

```{r sample-het}
het_check <- data.table::fread(
  file = "hapmap_ceu.het", 
  header = TRUE, 
  col.names = c("FID", "IID", "Ob_Hom", "Exp_Hom", "N_NM", "F_coef")
) %>% 
  mutate(het_rate = (N_NM - Ob_Hom) / N_NM)

het_sd_thres <- 3 # if sample's heterozygosity rate beyond 3 times the standard deviation of the average heterozygosity rate, considered as outliers. Can be changed.
het_upper <- mean(het_check$het_rate) + sd(het_check$het_rate) * het_sd_thres
het_lower <- mean(het_check$het_rate) - sd(het_check$het_rate) * het_sd_thres

## draw call rate against heterozygosity rate
inner_join(
  x = sample_miss,
  y = het_check, 
  by = "IID"
) %>% 
  mutate(
    lab = ifelse(het_rate > het_upper | het_rate < het_lower, IID, NA)
  ) %>% 
  ggplot(aes(x = het_rate, y = call_rate)) +
  geom_point(shape = 21) +
  geom_text(
    aes(label = lab), size = 3, hjust = -0.2, 
    na.rm = TRUE, show.legend = FALSE
  ) +
  geom_vline(
    xintercept = c(het_upper, het_lower), 
    linetype = 2, 
    color = "blue"
  ) +
  scale_x_continuous(labels = scales::percent_format(accuracy = 0.1)) +
  scale_y_continuous(labels = scales::percent_format(accuracy = 0.1)) +
  labs(x = "Heterozygosity Rate", y = "Call Rate", title = "Heterozygosity Check")

## store samples to be excluded due to aberrant heterozygosity rate
sample_to_exclude[["sample_het"]] <- filter(
  het_check, het_rate > het_upper | het_rate < het_lower
) %>% 
  select(FID, IID) %>% 
  mutate(QC = "Heterozygosity")

```

### Gender Check

The mean homozygosity rate across the X-chromosome was computed using the `--check-sex` command in PLINK.

Male samples are expected to have a homozygosity rate greater than 0.8, female samples are expected to have a homozygosity rate less than 0.2. 

```{r sample-gender-cmd}
system(command = "./plink --bfile hapmap_ceu --check-sex --out hapmap_ceu")
```

```{r sample-gender}
gender_check <- data.table::fread(
  file = "hapmap_ceu.sexcheck", 
  header = TRUE
)

inner_join(
  x = sample_miss,
  y = gender_check, 
  by = "IID"
) %>% 
  mutate(
    lab = ifelse(STATUS != "OK", IID, NA), 
    PEDSEX = ifelse(PEDSEX == 1, "Male", ifelse(PEDSEX == 2, "Female", NA))
  ) %>% 
  ggplot(aes(x = `F`, y = call_rate, color = PEDSEX)) +
  geom_point(shape = 21) +
  geom_text(aes(label = lab), size = 3, hjust = 1.2, na.rm = TRUE, show.legend = FALSE) +
  geom_vline(xintercept = c(0.2, 0.8), linetype = 2) +
  scale_x_continuous(breaks = seq(0, 1, 0.2)) +
  scale_y_continuous(labels = scales::percent_format(accuracy = 0.1)) +
  labs(x = "Homozygosity Rate", y = "Call Rate", title = "Gender Check")

## store samples to be excluded due to gender discrepancy
sample_to_exclude[["gender_dis"]] <- filter(gender_check, STATUS != "OK") %>% 
  select(FID, IID) %>% 
  mutate(QC = "Gender discrepancy")
```

**Quiz** 

* What's the possible reasons when a gender discrepancy is observed?

=> *e.g.*: sample contamination, simple entry error when recording phenotype.

### Relatedness Check

In order to identify pairs of samples who look too similar to each other, pairwise IBD (identity by descent) was computed, using the `--genome` command in PLINK.

Prior to the analysis, the dataset was pruned to exclude high-LD-regions, that is no variants with MAF < 0.01 and no pair of variants within 50 Kb have an r² > 0.2. (All these thresholds can be changed based on your prior knowledge of the data.)

If samples are not supposed to be related, the expected pairwise $\hat{\pi}$ ("PI_HAT") should be less than 0.2.

```{r sample-relatedness-cmd-0}
## prune high ld regions
system(command = "./plink --bfile hapmap_ceu --autosome --set-hh-missing --maf 0.01 --r2 --ld-window-r2 0.2 --ld-window-kb 50 --make-bed --out hapmap_ceu_pruned")

## make sure allelic frequency is representative after pruning
system(command = "./plink --bfile hapmap_ceu_pruned --freq --make-bed --out hapmap_ceu_pruned")

## estimate IBD
system(command = "./plink --bfile hapmap_ceu_pruned --read-freq hapmap_ceu_pruned.frq --genome --min 0.2 --out hapmap_ceu_pruned_0")

```

```{r sample-relatedness-0}
relatedness0 <- data.table::fread(
  file = "hapmap_ceu_pruned_0.genome", 
  header = TRUE
) 

data.frame(table(cut(
  relatedness0[["PI_HAT"]],
  breaks = c(0.2, 0.3, 0.4, 0.6, 0.8, 1),
  include.lowest = TRUE
))) %>%
  `colnames<-`(c("IBD windows", "Pair count")) %>%
  mutate(
    `IBD windows` = paste0(
      `IBD windows`,
      c(" = Second degree relatives", "", " = First degree relatives", "", " = MZ twins/duplicates")
    )
  )

```

The HapMap dataset is known to contain parent-offspring relations. Not surprisingly, 97 pairs of samples were identified as first-degree relatives.

Let's check how many pairs with PI_HAT > 0.2 after removing non-founders.

```{r sample-relatedness-cmd-1}
## exclude individuals with know parental ID
system(command = "./plink --bfile hapmap_ceu_pruned --filter-founders --make-bed --out hapmap_ceu_pruned_1")

## update allelic frequency
system(command = "./plink --bfile hapmap_ceu_pruned_1 --freq --make-bed --out hapmap_ceu_pruned_1")

## estimate IBD
system(command = "./plink --bfile hapmap_ceu_pruned_1 --read-freq hapmap_ceu_pruned_1.frq --genome --min 0.2 --out hapmap_ceu_pruned_1")

```

```{r sample-relatedness-1}
relatedness1 <- data.table::fread(
  file = "hapmap_ceu_pruned_1.genome", 
  header = TRUE
) 

data.frame(table(cut(
  relatedness1[["PI_HAT"]],
  breaks = c(0.2, 0.3, 0.4, 0.6, 0.8, 1),
  include.lowest = TRUE
))) %>%
  `colnames<-`(c("IBD windows", "Pair count")) %>%
  mutate(
    `IBD windows` = paste0(
      `IBD windows`,
      c(" = Second degree relatives", "", " = First degree relatives", "", " = MZ twins/duplicates")
    )
  ) 

```

There still one pair of samples has PI_HAT > 0.2, we'll keep the individual with higher call rate for each pair of related samples.

```{r sample-relatedness-2}
related_samples <- select(relatedness1, FID1, IID1, FID2, IID2, PI_HAT) %>% 
  left_join(
    select(sample_miss, FID1 = FID, IID1 = IID, F_MISS1 = F_MISS),
    by = c("FID1", "IID1")
  ) %>% 
  left_join(
    select(sample_miss, FID2 = FID, IID2 = IID, F_MISS2 = F_MISS),
    by = c("FID2", "IID2")
  ) %>% 
  mutate(
    call_rate1 = 1 - F_MISS1, 
    call_rate2 = 1 - F_MISS2
  ) %>% 
  arrange(desc(PI_HAT))

select(related_samples, -contains("MISS"))

## store samples to be excluded due to relatedness
sample_to_exclude[["sample_relatedness"]] <- filter(related_samples, PI_HAT > 0.2) %>% 
  select(matches("FID|IID|call_rate")) %>% 
  unite(col = FID_IID1, FID1, IID1, sep = "_-_") %>% 
  unite(col = FID_IID2, FID2, IID2, sep = "_-_") %>% 
  mutate(FID_IID = ifelse(call_rate1 >= call_rate2, FID_IID2, FID_IID1)) %>% 
  separate(col = "FID_IID", into = c("FID", "IID"), sep = "_-_") %>% 
  distinct(FID, IID) %>% 
  mutate(QC = "Relatedness")

```

### Population Structure

The `hapmap_ceu` dataset was pruned using the following cut-offs:

* variants call rate > 90%
* MAF ≥ 0.01
* HWE > 1e-04
* samples call rate > 99%
* heterozygosity rate within 3 times SD from the mean heterozygosity rate

(All these thresholds can be changed)

```{r sample-pop-1}
## remove bad quality samples for PCA
rbind(
  sample_to_exclude[["sample_cr"]],
  sample_to_exclude[["sample_het"]]
) %>% 
  distinct(FID, IID) %>% 
  data.table::fwrite(
    file = "samples_toexclude_pca.txt",
    row.names = FALSE, 
    col.names = FALSE, 
    quote = FALSE, 
    sep = " "
  )

```

```{r sample-pop-hapmap-cmd}
## prune hapmap_ceu data for PCA
system(command = "./plink --bfile hapmap_ceu --remove samples_toexclude_pca.txt --autosome --snps-only --maf 0.01 --hwe 0.0001 --geno 0.1 --make-bed --out hapmap_ceu_pca")
```

```{r sample-pop-2}
## extract hapmap_ceu SNPs
data.table::fwrite(
  data.table::fread("hapmap_ceu_pca.bim", header = FALSE)[, "V2"], 
  file = "hapmap_ceu_pca_snps.txt",
  sep = "\n", 
  row.names = FALSE, 
  col.names = FALSE, 
  quote = FALSE
)

## number of samples and variants ready to be merged with 1KG data
hapmap_sample_pca <- scales::comma(as.numeric(
  gsub(
    pattern = "\\s+([0-9]+)\\s.*", 
    replacement = "\\1", 
    x = system("wc -l hapmap_ceu_pca.fam", intern = TRUE)
  )
))

hapmap_variant_pca <- scales::comma(as.numeric(
  gsub(
    pattern = "\\s+([0-9]+)\\s.*", 
    replacement = "\\1", 
    x = system("wc -l hapmap_ceu_pca.bim", intern = TRUE)
  )
))

```

After prunning, it remains `r hapmap_sample_pca` samples and `r hapmap_variant_pca` variants to be merged with 1,000 Genomes Project public data.

**[DO NOT RUN FOLLOWING SECTIONS !]**

**Attention:** huge 1,000 Genomes Project data with >60G, please do not download if your disk memory is not allowed, download time is about 2 hours.

```{r get-merge-1kg, eval=FALSE}
#######
# Do not run
#######

### get 1kg data under build 36
system(command = "wget ftp://ftp-trace.ncbi.nih.gov/1000genomes/ftp/release/20100804/20100804.ALL.panel") # 1kg phenotype
system(command = "wget ftp://ftp-trace.ncbi.nih.gov/1000genomes/ftp/release/20100804/ALL.2of4intersection.20100804.genotypes.vcf.gz") # 1kg genotype
system(command = "./plink --vcf ALL.2of4intersection.20100804.genotypes.vcf.gz --make-bed --out ALL.2of4intersection.20100804.genotypes") # transform VCF into plink format
system(command = "./plink --bfile ALL.2of4intersection.20100804.genotypes --set-missing-var-ids @:#[b37]\$1,\$2 --make-bed --out ALL.2of4intersection.20100804.genotypes_no_missing_IDs") # optional, naming missing variant ID to chr:pos[b37]A1,A2, eg: chr1:10583[b37]G,A
system(command = "./plink --bfile ALL.2of4intersection.20100804.genotypes_no_missing_IDs --extract hapmap_ceu_pca_snps.txt --snps-only --make-bed --out 1kg_small") # extract variants present in hapmap_ceu dataset


### QC 1kg data
system(command = "./plink --bfile 1kg_small --mind 0.1 --make-bed --out 1kg_small
./plink --bfile 1kg_small --geno 0.1 --maf 0.01 --hwe 0.0001 --make-bed --out 1kg_small")

## extract 1kg variant from hapmap_ceu data
system(command = "awk '{print$2}' 1kg_small.bim > 1kg_pca_snps.txt")
system(command = "./plink --bfile hapmap_ceu_pca --extract 1kg_pca_snps.txt --recode --make-bed --out hapmap_ceu_pca_1kg")


### merge
## update 1kg base-pair positions based on hapmap_ceu
system(command = "awk '{print$2,$4}' hapmap_ceu_pca_1kg.map > hapmap_build.txt")
system(command = "./plink --bfile 1kg_small --update-map hapmap_build.txt --make-bed --out 1kg_small_clean")

## make sure reference allele is the same for variants
system(command = "awk '{print$2,$5}' 1kg_small_clean.bim > 1kg_ref-list.txt")
system(command = "./plink --bfile hapmap_ceu_pca_1kg --reference-allele 1kg_ref-list.txt --make-bed --out hapmap_ceu_pca_1kg_adj")

## resolve strand issues
system(command = "./plink --bfile 1kg_small_clean --bmerge hapmap_ceu_pca_1kg_adj.bed hapmap_ceu_pca_1kg_adj.bim hapmap_ceu_pca_1kg_adj.fam --make-bed --out hapmap_1kg")

## error when matching, flip variant strand
system(command = "./plink --bfile 1kg_small_clean --flip hapmap_1kg-merge.missnp --make-bed --out 1kg_small_flipped")

## try to merge again
system(command = "./plink --bfile 1kg_small_flipped --bmerge hapmap_ceu_pca_1kg_adj.bed hapmap_ceu_pca_1kg_adj.bim hapmap_ceu_pca_1kg_adj.fam --make-bed --out hapmap_1kg")

## still error, exclude mis-matching variants in both datasets
system(command = "./plink --bfile 1kg_small_flipped --exclude hapmap_1kg-merge.missnp --make-bed --out 1kg_small_flipped_clean")
system(command = "./plink --bfile hapmap_ceu_pca_1kg_adj --exclude hapmap_1kg-merge.missnp --make-bed --out hapmap_ceu_pca_1kg_adj_clean")

## merge 1kg with hapmap_ceu data
system(command = "./plink --bfile 1kg_small_flipped_clean --bmerge hapmap_ceu_pca_1kg_adj_clean.bed hapmap_ceu_pca_1kg_adj_clean.bim hapmap_ceu_pca_1kg_adj_clean.fam --make-bed --out hapmap_1kg")

```

```{r sample-pop-3, eval=FALSE}
#######
# Do not run
#######

### extract SNPs in common
intersect(
  data.table::fread("1kg_small_flipped_clean.bim", header = FALSE)[["V2"]],
  data.table::fread("hapmap_ceu_pca_1kg_adj_clean.bim", header = FALSE)[["V2"]]
) %>%
  data.table::fwrite(
    file = "hapmap_1kg_snps.txt",
    col.names = FALSE, 
    row.names = FALSE, 
    quote = FALSE, 
    sep = "\n"
  )
```

```{r eval=FALSE}
#######
# Do not run
#######

### extract SNPs in common from merged dataset
system(command = "./plink --bfile hapmap_1kg --extract hapmap_1kg_snps.txt --make-bed --out hapmap_1kg")
```

```{r sample-pop-4, eval=FALSE}

### check palindromic variants
data.table::fread("hapmap_1kg.bim", header = FALSE) %>%
  unite(alleles, V5, V6, remove = FALSE, sep = "") %>%
  filter(alleles %in% c("GC", "CG", "AT", "TA")) %>%
  select(V2) %>%
  data.table::fwrite(
    file = "hapmap_1kg_palindromic.txt",
    row.names = FALSE, 
    col.names = FALSE, 
    sep = "\n",
    quote = FALSE
  )

```

```{r eval=FALSE}
#######
# Do not run
#######

### remove palindromic variants
system(command = "./plink --bfile hapmap_1kg --exclude hapmap_1kg_palindromic.txt --make-bed --out hapmap_1kg")

### run PCA
system(command = "./plink --bfile hapmap_1kg --make-grm-bin --pca 10 --thread-num 10 --out hapmap_1kg") # calculate only the first 10 PCs using 10 cores

```

After running PCA, you will get two files, one for the eigenvalues (`.eigenval`) and another for the eigenvectors (`.eigenvec`). These two files are prepared in advance, named `hapmap_1kg.eigenval` and `hapmap_1kg.eigenvec`. Eigenvector contains the 10 first principal components. Now, let's visualize the population structure.

**Quiz** 

* Which 1,000 genomes population is the closest to HapMap population? What is your conclusion?

```{r sample-pop-pca}
### merge phenotype with eigenvectors
data_pca <- data.table::fread("hapmap_1kg.fam", header = FALSE) %>% 
  select(IID = V2, pheno = V5) %>% 
  left_join(
    rbind(
      data.table::fread("relationships_w_pops_041510.txt", header = TRUE) %>% 
        filter(population == "CEU") %>% 
        select(FID, IID) %>% 
        mutate(population = "HapMap"), 
      read.table("20100804.ALL.panel", header = FALSE, fill = TRUE) %>% 
        select(FID = V1, IID = V1, population = V2)
    ) %>% 
      distinct(IID, .keep_all = TRUE), 
    by = "IID"
  ) %>% 
  left_join(
    data.table::fread("hapmap_1kg.eigenvec", header = FALSE), 
    by = c("FID" = "V1", "IID" = "V2")
  ) %>% 
  left_join(
    data.table::fread("integrated_call_samples_v3.20130502.ALL.panel", header = TRUE, fill = TRUE) %>% 
      distinct(pop, super_pop), 
    by = c("population" = "pop")
  ) %>% 
  mutate(
    super_pop = ifelse(population == "HapMap", "HapMap", super_pop),
    super_pop = relevel(factor(super_pop), ref = "HapMap")
  ) %>% 
  rename(x = V3, y = V4) %>% 
  left_join(
    y = stats::aggregate(cbind(x, y) ~ super_pop, data = ., mean) %>% # calculate centroid of each super population
      `colnames<-`(paste0(colnames(.), c("", ".centroid", ".centroid"))),
    by = "super_pop"
  )

### get the centroid of each super population in 1KG
centre_pca <- data_pca %>%
  filter(super_pop != "HapMap") %>%
  select(IID, super_pop, contains("centroid")) %>%
  mutate_at(.vars = vars(IID, super_pop), .funs = as.character)

### check which is the closest 1kg super population
check_ethnicity <- data.table::fread("hapmap_ceu_pca_1kg_adj_clean.fam", header = FALSE) %>%
  select(FID = V1, IID = V2) %>%
  mutate_all(as.character) %>%
  left_join(
    filter(data_pca, super_pop == "HapMap"),
    by = c("FID", "IID")
  ) %>%
  rowwise() %>%
  mutate(
    ClosestPop = centre_pca[["super_pop"]][which.min(sqrt(
      (x - centre_pca[["x.centroid"]])^2 + (y - centre_pca[["y.centroid"]])^2
    ))]
  ) %>% 
  ungroup()

unique(check_ethnicity$ClosestPop[which(check_ethnicity$super_pop == "HapMap")])
```

```{r}
### draw the first factorial plane
ggplot(
  data = data_pca, aes(x = x, y = y, color = super_pop)
) +
  geom_hline(yintercept = 0, linetype = 2) +
  geom_vline(xintercept = 0, linetype = 2) +
  geom_point(na.rm = TRUE, shape = 21) +
  scale_color_viridis_d(direction = -1) +
  labs(
    x = "First Principal Component", 
    y = "Second Principal Component", 
    colour = NULL, 
    title = "HapMap vs. 1,000 Genomes"
  ) +
  theme(
    legend.position = "bottom", 
    legend.key.size = unit(0.5, units = "lines")
  )

## variation explained by first 10 PCs
inertia_pca <- data.table::fread("hapmap_1kg.eigenval", header = FALSE) %>%
  mutate(y = V1 / sum(V1))

inertia_pca %>%
  mutate(x = seq_along(V1)) %>%
  ggplot(mapping = aes(x = x, y = y)) +
  geom_bar(stat = "identity", fill = scales::viridis_pal(begin = 0.5, end = 0.5)(1)) +
  labs(y = "Inertia", x = "Principal Components", title = "PCA Scree Plot") +
  scale_x_continuous(breaks = seq_len(10)) +
  scale_y_continuous(label = scales::percent_format(), expand = expansion(mult = c(0, 0.05)))

```

### Final Sample-based QC

According to previous steps, now we know which samples have bad quality and should be excluded for variant-based QC. 

**Quiz**

* How many samples were excluded? 

```{r sample-exclusion}
sample_excluded <- do.call("rbind", sample_to_exclude) %>% 
  distinct(FID, IID)
nrow(sample_excluded)

data.table::fwrite(
  sample_excluded,
  file = "samples_excluded.txt",
  sep = "\t", 
  col.names = TRUE, 
  row.names = FALSE,
  quote = FALSE
)
  
```

* Do you think sample with gender discrepancy should be removed?

=> if it's due to sample contamination, it should be excluded before downstream analysis; if it's due to phenotype file error, we can just correct the phenotype file and keep this sample in analysis.

```{r sample-exclusion-cmd}
## final sample-based QC
system(command = "./plink --bfile hapmap_ceu --filter-founders --remove samples_excluded.txt --snps-only --set-hh-missing --make-bed --out hapmap_ceu_good_samples")
```

## Variant-based QC

### Variant Call Rate

Proportion of missing genotypes per variant was estimated using the `--missing` command in PLINK. Call rate was calculated as 1 - missing rate.

```{r variant-cr-cmd}
system(command = "./plink --bfile hapmap_ceu_good_samples --missing --out hapmap_ceu_good_samples")
```

```{r variant-cr}
variant_miss <- data.table::fread(
  file = "hapmap_ceu_good_samples.lmiss", 
  header = TRUE
) %>% 
  mutate(call_rate = 1 - F_MISS)

variant_callrate_thres <- 0.95 # threshold can be changed

## visualize call rate per variant
arrange(variant_miss, call_rate) %>% 
  ggplot(aes(x = seq_along(SNP), y = call_rate)) +
  geom_point(shape = 21) +
  geom_hline(yintercept = variant_callrate_thres, color = "blue", linetype = 2) +
  scale_x_continuous(labels = scales::comma_format()) +
  scale_y_continuous(labels = scales::percent_format(accuracy = 0.1)) +
  labs(x = "Variants", y = "Call Rate", title = "Variant Call Rate Check") +
  theme(axis.text.x = element_text(hjust = 1, angle = 10))

```

### Minor Allele Frequency Distribution

The MAF were computed using the `--freq` command in PLINK.

```{r variant-maf-cmd}
system(command = "./plink --bfile hapmap_ceu_good_samples --freq --out hapmap_ceu_good_samples")
```

```{r variant-maf}
## visualize the allelic frequency distribution
data.table::fread(
  file = "hapmap_ceu_good_samples.frq", 
  header = TRUE
) %>% 
  mutate(
    maf_dist = cut(MAF, breaks = c(0, 0.01, 0.05, 0.1, 0.2, 0.3, 0.4, 0.5), include.lowest = TRUE)
  ) %>% 
  drop_na(MAF) %>% 
  ggplot(aes(x = maf_dist)) +
  geom_bar(fill = scales::viridis_pal(begin = 0.5, end = 0.5)(1)) +
  labs(x = "MAF", y = "Number of variants", title = "MAF Distribution") +
  scale_y_continuous(labels = scales::comma_format(), expand = expansion(mult = c(0, 0.1)))

```

### Hardy-Weinberg Equilibrium

Hardy-Weinberg equilibrium (HWE) p-value was computed using the `--hardy` command in PLINK.

```{r variant-hwe-cmd}
system(command = "./plink --bfile hapmap_ceu_good_samples --hardy --out hapmap_ceu_good_samples")
```


```{r variant-hwe}
hwe <- data.table::fread(
  file = "hapmap_ceu_good_samples.hwe", 
  header = TRUE
)

hwe_p_thres <- 0.0001 # threshold can be changed, variants below this threshold will be excluded

## visualize variants with hwe p < 1e-3
hwe_significant <- filter(hwe, TEST == "ALL", P < hwe_p_thres * 10) %>% 
  arrange(P)
ggplot(data = hwe_significant, aes(x = seq_along(SNP), y = P)) +
  geom_point(shape = 21) +
  geom_hline(yintercept = hwe_p_thres, linetype = 2, color = "blue") +
  scale_y_continuous(trans = pval_trans()) +
  labs(
    x = "Variants", y = "HWE P-value", 
    title = "HWE P-value Distribution", 
    subtitle = paste0("(truncated at p < ", hwe_p_thres * 10, ")")
  )
  
```

## Final Data Management At One Time

```{r final-cmd}
system(command = paste(
  "./plink --bfile hapmap_ceu", 
  "--filter-founders",
  "--remove samples_excluded.txt",
  "--snps-only",
  "--set-hh-missing",
  "--geno 0.05",
  "--hwe 0.0001",
  "--recode",
  "--make-bed",
  "--out hapmap_ceu_clean"
))
```

# Association Test

We will run association test only for chromosome 22 using logistic regression model with the `--logistic` command in PLINK.

Optional: use `--covar` command to indicate the files containing covariables, use `--covar-name` to specify a subset of covariable by column name in covariable file. 

Here, the first two columns of the covariable file `covars.txt` are family ID (FID) and sample ID (IID). The rest of columns are sex and the first 2 PCs, which will used as covariables in regression model.

```{r eval=FALSE}
## extract covariables
left_join(
  data.table::fread("hapmap_ceu_clean.fam", header = FALSE) %>% 
    mutate(V1 = as.character(V1)) %>% 
    select(V1, V2, sex = V5),
  data.table::fread("hapmap_1kg.eigenvec", header = FALSE) %>% 
    select(V1, V2, V3, V4), 
  by = c("V1", "V2")
) %>% 
  rename(FID = V1, IID = V2, PC1 = V3, PC2 = V4) %>% 
  data.table::fwrite(
    file = "covars.txt", 
    col.names = TRUE,
    row.names = FALSE, 
    sep = "\t", 
    quote = FALSE
  )

```

```{r run-association-test}
### logistic regression with covariables
system(command = "./plink --bfile hapmap_ceu_clean --chr 22 --logistic --covar covars.txt --covar-name sex, PC1, PC2 --out res_glm_covar_chr22")
```

Detail description of association analysis' output: https://zzz.bwh.harvard.edu/plink/anal.shtml

```{r}
## read regression results
res_glm_covar <- data.table::fread(
  "res_glm_covar_chr22.assoc.logistic", header = TRUE
)
res_glm_covar
```

If covariables are included in model, each variant will have the same number of rows as the number of covariables + 1.

The `TEST` column indicates the tested variables. `ADD` corresponds to the tested variant under additive model. When there are NAs, it might due to all samples carry the same genotype for this variant, or the tested allele was not carried in samples (*e.g.*: in reference genome is A/G, but samples are A/T), the estimation was not completed.

# Visualization

```{r}
## keep only rows of variants' estimation
res_glm_covar <- filter(res_glm_covar, TEST == "ADD", ! is.na(P)) %>% 
  mutate(FDR = p.adjust(P, method = "BH")) %>% 
  arrange(FDR)

head(res_glm_covar)

```

## Quantile-Quantile Plot

QQ plot draw the observed p-values from regression model versus the expected p-values of doing the same number of tests in a random dataset.

In a GWAS, we expect to see only a few (some times none) variants are associated with the trait, so nearly all the p-values should follow the uniform distribution.

When a large proportion of p-values depart away from the expected uniform distribution (points not line along the diagonal line), it usually indicates either poor data quality or population substructure not included in model.

The genomic inflation factor $\lambda$ is a numerical summary of the departure from the uniform distribution, we expect to have a $\lambda$ close to 1.

$$
\lambda = \frac{median(\chi^2_{obs})}{median(\chi^2_{exp})} = \frac{median(qnorm(p/2)^2)}{0.456}
$$

```{r}
ggqqplot(res_glm_covar[, "P"])
```

Following the QQ plot for genome-wide tested variants.

![](figs/qq_plot.png){width=80%}

## Manhattan Plot

Manhattan plot was used to show p-values of all studied variants against their position in chromosomes.

```{r}
draw_manhattan(
  data = res_glm_covar,
  x = "BP",
  y = "P",
  chr = "CHR",
  label_y = "P-Value",
  alpha = 0.05
) +
  scale_x_continuous(label = "22", breaks = 1)
```

Following the manhattan plot for genome-wide tested variants.

![](figs/manhattan_plot.png)

# R Session Information

```{r session-info, echo=FALSE, results="markup"}
options("width" = 110)
sessionInfo()
```

